### PR TITLE
Add more tests for sensor

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
Failed to run the sensor in the agent due to following error.
```bash
 {"asctime": "2023-10-14 10:58:59,300", "name": "flytekit", "levelname": "ERROR", "message": "task_type: \"sensor\"\nresource_meta: \"\\200\\005\\225l\\000\\000\\000\\000\\000\\000\\000}\\224(\\214\\rsensor_module\\224\\214\\033flytekit.sensor.file_sensor\\224\\214\\013sensor_name\\224\\214\\nFileSensor\\224\\214\\006inputs\\224}\\224\\214\\004path\\224\\214\\010/tmp/123\\224su.\"\n"}
 {"asctime": "2023-10-14 10:58:59,303", "name": "flytekit", "levelname": "ERROR", "message": "Importing module usr.local.bin.pyflyte from file /usr/local/bin/pyflyte"}
```
Previously, it worked, but this [PR](https://github.com/flyteorg/flytekit/commit/b072130a6edbf80cf5406d35c6f8a3a7a60fefbe) breaks it.

We shouldn't import usr.local.bin.pyflyte since it's not even a module.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
I've tested it in the flytectl demo cluster

## Tracking Issue
https://github.com/flyteorg/flyte/issues/3936
https://github.com/flyteorg/flyte/issues/4245

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by wild-endeavor** [comment]
@pingsutw can you point me to a minimal repro for this?  want to dig a bit deeper to understand what's happening.

**Comment by pingsutw** [comment]
1. Run a agent server pyflyte serve
2. Send a gRPC request to the server by running below script
```python
import threading

import grpc
from flyteidl.service.agent_pb2_grpc import AsyncAgentServiceStub
from flyteidl.admin import agent_pb2
from matplotlib import pyplot as plt
from timeit import default_timer as timer
import tempfile
from unittest.mock import MagicMock

import grpc
import pytest
import cloudpickle
from flyteidl.admin.agent_pb2 import (
    SUCCEEDED,
    DeleteTaskResponse,
)

import flytekit.models.interface as interface_models
from flytekit.extend.backend.base_agent import (
    AgentRegistry,
)
from flytekit.models import literals, types
from flytekit.sensor import FileSensor
from flytekit.sensor.base_sensor import SENSOR_MODULE, SENSOR_NAME
from tests.flytekit.unit.extend.test_agent import get_task_template

channel = grpc.insecure_channel('localhost:8000')
stub = AsyncAgentServiceStub(channel)


interfaces = interface_models.TypedInterface(
        {
            "path": interface_models.Variable(types.LiteralType(types.SimpleType.STRING), "description1"),
        },
        {},
    )
tmp = get_task_template("sensor")
tmp._custom = {
    SENSOR_MODULE: FileSensor.__module__,
    SENSOR_NAME: FileSensor.__name__,
}


file = tempfile.NamedTemporaryFile()

tmp._interface = interfaces

task_inputs = literals.LiteralMap(
    {
        "path": literals.Literal(scalar=literals.Scalar(primitive=literals.Primitive(string_value=file.name))),
    },
)

if __name__ == '__main__':
    try:
        for i in range(1):
            res = stub.CreateTask(request=agent_pb2.CreateTaskRequest(template=tmp.to_flyte_idl(), inputs=task_inputs.to_flyte_idl(), output_prefix="/tmp"))
            res = stub.GetTask(request=agent_pb2.GetTaskRequest(task_type="sensor", resource_meta=res.resource_meta))
            print(res)
    except Exception as e:
        print("error:", e)

```

**Comment by pingsutw** [comment]
or register a dynamic task, and run it.
```python
import sys
from typing import List
from flytekit import task, workflow, dynamic, Resources
from flytekit.remote import *
from flytekit.configuration import Config

@task(requests=Resources(mem="500Mi"))
def add1(x: int) -> int:
    return x + 1

@dynamic(requests=Resources(mem="500Mi"))
def q(n: int) -> int:
    ctr = 0
    for _ in range(n):
        ctr = add1(x=ctr)
    return ctr

@task
def fin(x: int):
    print(x)

@workflow
def workflow():
    fin(x=q(n=5))

```

**Comment by pingsutw** [comment]
Find the same issue in the map task. you can reproduce it by running below workflow
```python
import typing

import pandas as pd
from click.testing import CliRunner

from flytekit import task, workflow, ImageSpec, map_task
from flytekit.clis.sdk_in_container import pyflyte
from flytekit.deck.renderer import TopFrameRenderer
from typing_extensions import Annotated


new_flytekit = "git+https://github.com/flyteorg/flytekit.git@542d63b0de22ad8de77bd20d4d3e7e246f7ced37"

image_spec = ImageSpec(base_image="python:3.8-slim-buster", packages=[new_flytekit, "pandas", "pillow"], apt_packages=["git"], registry="pingsutw", name="flytekit1")


@task(enable_deck=True, container_image=image_spec)
def a_mappable_task(a: int) -> str:
    inc = a + 2
    stringified = str(inc)
    return stringified


@task(enable_deck=True, container_image=image_spec)
def t1() -> (Annotated[pd.DataFrame, TopFrameRenderer()], int):
    return pd.DataFrame({"col1": [1, 2, 3], "col2": list("abc")}), 3


@workflow
def wf(a: typing.List[int] = [1, 2, 3]):
    t1()
    map_task(a_mappable_task)(a=a)


if __name__ == '__main__':
    runner = CliRunner()
    result = runner.invoke(pyflyte.main, ["run", "--remote", "flyte-example/deck_example.py", "wf"])
    print(result.output)

```

**Comment by pingsutw** [comment]
image

File: flytekit/core/tracker.py [hunk 1]
```diff
@@ -64,6 +64,8 @@ def _find_instance_module():
             if frame.f_code.co_name == "<module>" and "__name__" in frame.f_globals:
                 if frame.f_globals["__name__"] != "__main__":
                     return frame.f_globals["__name__"], frame.f_globals["__file__"]
+                if frame.f_globals["__file__"].endswith("pyflyte"):
+                    return None, None
                 # if the remote_deploy command is invoked in the same module as where
                 # the app is defined, get the module from the file name
                 mod = InstanceTrackingMeta._get_module_from_main(frame.f_globals)
```
Comment (pingsutw) [inline flytekit/core/tracker.py:71]: cc @cosmicBboy is there way to `get_module_from_main` only when running eager workflow?

File: flytekit/core/tracker.py [hunk 2]
```diff
@@ -64,6 +64,8 @@ def _find_instance_module():
             if frame.f_code.co_name == "<module>" and "__name__" in frame.f_globals:
                 if frame.f_globals["__name__"] != "__main__":
                     return frame.f_globals["__name__"], frame.f_globals["__file__"]
+                if frame.f_globals["__file__"].endswith("pyflyte"):
```
Comment (cosmicBboy) [inline flytekit/core/tracker.py:67]: let's create a tuple of the executables we want to exclude.

Comment (pingsutw) [reply flytekit/core/tracker.py]: I just removed the changes from `tracker.py`. Now, This PR only adds the tests for the agent.

**Comment by pingsutw** [comment]
note: we have fixed issue in https://github.com/flyteorg/flytekit/pull/1902

**Review by eapolinario — APPROVED** [review]
thank you. This will help a lot in the followup PR.

**Review by wild-endeavor — APPROVED** [review]
